CommonMark/Parser/TableMethods.cs

```diff
@@ -0,0 +1,365 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using CommonMark.Syntax;
+
+namespace CommonMark.Parser
+{
+	internal static class TableMethods
+	{
+		static List<string> ParseTableLine(StringPart part, StringBuilder sb)
+		{
+			if (part.Source == null)
+				return null;
+
+			string line = part.Source.Substring(part.StartIndex, part.Length);
+			line = line.TrimEnd('\n');
+
+			var ret = new List<string>();
+
+			var i = 0;
+
+			if (i < line.Length && line[i] == '|') i++;
+
+			while (i < line.Length && char.IsWhiteSpace(line[i])) i++;
+
+			for (; i < line.Length; i++)
+			{
+				var c = line[i];
+				if (c == '\\')
+				{
+					i++;
+					if (i < line.Length && line[i] == '|')
+					{
+						sb.Append(line[i]);
+						continue;
+					}
+					i--;
+				}
+
+				if (c == '|')
+				{
+					ret.Add(sb.ToString());
+					Utilities.ClearStringBuilder(sb);
+				}
+				else
+				{
+					sb.Append(c);
+				}
+			}
+
+			if (sb.Length != 0)
+			{
+				ret.Add(sb.ToString());
+				Utilities.ClearStringBuilder(sb);
+			}
+
+			return ret;
+		}
+
+		static void MakeTableCells(Block row, StringBuilder sb)
+		{
+			var offset = 0;
+
+			var parts = row.StringContent.RetrieveParts();
+			foreach (var part in parts.Array)
+			{
+				if (part.Length <= 0)
+					continue;
+
+				string asStr = part.Source.Substring(part.StartIndex, part.Length);
+
+				for (var i = 0; i < asStr.Length; i++)
+				{
+					var c = asStr[i];
+
+					if (c == '|' && (i == 0 || !Utilities.IsInlineSpanSymbol(asStr[i - 1])))
+					{
+						var text = sb.ToString();
+						Utilities.ClearStringBuilder(sb);
+
+						if (text.Length > 0)
+							MakeCell(text, row, ref offset);
+
+						offset += text.Length;
+
+						// skip the |
+						offset++;
+						continue;
+					}
+
+					if (c == '\\')
+					{
+						sb.Append(c);
+						if (i + 1 < asStr.Length)
+						{
+							if (Utilities.IsEscapableSymbol(asStr[i + 1]))
+								sb.Append(asStr[i + 1]);
+						}
+						i++;
+					}
+					else
+					{
+						sb.Append(c);
+					}
+				}
+			}
+
+			if (sb.Length > 0)
+			{
+				var text = sb.ToString();
+				Utilities.ClearStringBuilder(sb);
+
+				if (text.Length > 0)
+					MakeCell(text, row, ref offset);
+			}
+		}
+
+		private static void MakeCell(string text, Block row, ref int offset)
+		{
+			int length = text.Length;
+			string trimmedText = text.TrimStart();
```

Comment:
a new string copy `trimmedText` should not be created - instead the whitespaces should be counted within the `text` and then the StringContent created from the `text` instance.

Though as before, it could be left as a `//TODO:`